### PR TITLE
Tab completion for share namespaces

### PR DESCRIPTION
**Pull request by ChrisPenner**
## Overview

Provides tab-completion for share users and namespaces for `push` and `pull`.

## Implementation notes

* Calls out to the `search` and `namespace browse` apis for share when tab is hit to find completions.
* Currently doesn't do any form of caching since the results could change on the remote at any time

## Test coverage

* I added a transcript for this for testing, but I suspect I probably shouldn't check it in, since we may not want transcripts which perform http requests. Let me know what you think, but I'm leaning towards deleting it.

## Open questions to answer before merging

* This causes the Unison CLI to depend on the Share API interface, is that acceptable?
* Should we move more of these types from Enlil into the `unison-share-api` package to avoid them going out of sync and to get better type guarantees? Would it be a pain to develop against them in Enlil if we were to do that?

- - -
This change is [Reviewable](https://reviewable.io/reviews/unisonweb/unison/3426)

### CODE REVIEW

File: unison-cli/src/Unison/CommandLine/Completion.hs [hunk 1]
```diff
@@ -339,3 +354,101 @@ fixupCompletion q cs@(h : t) =
...
+    req <- MaybeT $ pure (HTTP.requestFromURI uri)
+    fullResp <- liftIO $ UnliftIO.tryAny $ HTTP.httpLbs req httpManager
+    resp <- either (const empty) pure $ fullResp
+    results <- (MaybeT . pure . Aeson.decode @[SearchResult] $ HTTP.responseBody resp)
+    pure $
+      results
+        & filter (\SearchResult {tag} -> tag == "User")
+        & fmap handle
+
+data SearchResult = SearchResult
```
Comment (ChrisPenner) [inline unison-cli/src/Unison/CommandLine/Completion.hs:444]: Here's a data-type I copied over from Enlil, should we just move it to the `unison-share-api` package instead so that Share and unison-cli are in sync?

File: unison-share-api/src/Unison/Server/Types.hs [hunk 2]
```diff
@@ -239,21 +248,44 @@ data NamedType = NamedType
 instance ToJSON NamedType where
   toEncoding = genericToEncoding defaultOptions
 
+instance FromJSON NamedType where
+  parseJSON = genericParseJSON defaultOptions
```
Comment (ChrisPenner) [inline unison-share-api/src/Unison/Server/Types.hs:252]: I really don't like using "generic" JSON instances, they're too susceptible to change upon refactoring or even renaming fields; but I used it here because it has to match the `ToJSON` which is currently using the generic method.

File: unison-share-api/src/Unison/Server/Types.hs [hunk 3]
```diff
@@ -239,21 +248,44 @@ data NamedType = NamedType
...
+
 deriving instance ToSchema NamedType
 
 instance ToJSON TermTag where
   toJSON = \case
-    Doc -> "doc"
-    Test -> "test"
-    Plain -> "plain"
+    Doc -> "Doc"
```
Comment (ChrisPenner) [inline unison-share-api/src/Unison/Server/Types.hs]: I believe I mistakenly changed the JSON serialization of these tags in a previous commit which hasn't been deployed yet 😬 

This should change them back.

File: unison-src/transcripts/tab-completion.md [hunk 4]
```diff
@@ -50,3 +50,9 @@ unique type subnamespace.AType = A | B
 .> debug.tab-complete cd subnamespace
 .> debug.tab-complete cd subnamespace.
 ```
+
+## Tab complete against Share
+
+```ucm
+.> debug.tab-complete pull unison.pub
```
Comment (ChrisPenner) [inline unison-src/transcripts/tab-completion.md]: Here's the transcript I'm thinking of just removing, since it does a web-request in the transcript, and isn't deterministic because it relies on the state of Share.

File: unison-share-api/src/Unison/Server/Types.hs (hunk 2 above)
Comment (aryairani) [reply unison-share-api/src/Unison/Server/Types.hs:252]: We can make them both explicit instead

File: unison-src/transcripts/tab-completion.md (hunk 4 above)
Comment (aryairani) [reply unison-src/transcripts/tab-completion.md]: I think we might collect more of these "transcripts we want to run but don't want to run", maybe we should collect them in some directory and figure out what to do with them later, rather than just deleting? But no strong feelings.

File: unison-share-api/src/Unison/Server/Types.hs (hunk 2 above)
Comment (ChrisPenner) [reply unison-share-api/src/Unison/Server/Types.hs:252]: We can, it's just a bit annoying to figure out how to match it exactly to what it's currently doing (not that hard, just a bit annoying and making more work for me).

I'd rather go through and do them all at once at some point.

Comment (aryairani) [reply unison-share-api/src/Unison/Server/Types.hs:252]: Yeah that's fine.

File: unison-cli/src/Unison/Auth/Tokens.hs [hunk 5]
```diff
@@ -54,7 +52,7 @@ performTokenRefresh :: MonadIO m => URI -> Tokens -> m (Either CredentialFailure
 performTokenRefresh discoveryURI (Tokens {refreshToken = currentRefreshToken}) = runExceptT $
   case currentRefreshToken of
     Nothing ->
-      throwError $ (RefreshFailure . Text.pack $ "Unable to refresh authentication, please run " <> patternName IP.authLogin <> " and try again.")
+      throwError $ (RefreshFailure . Text.pack $ "Unable to refresh authentication, please run auth.login and try again.")
```
Comment (aryairani) [inline unison-cli/src/Unison/Auth/Tokens.hs:55]: Why this change?

**Review by aryairani — APPROVED** [review]

**Comment by hojberg** [comment]
This is really cool! How does it feel to use? How's the delay?

File: unison-cli/src/Unison/Auth/Tokens.hs (hunk 5 above)
Comment (ChrisPenner) [reply unison-cli/src/Unison/Auth/Tokens.hs:55]: Circular imports :'( 

Now that the CLI depends on the auth client, the auth client can't depend on the Input Patterns.

I'd rather not do this of course, but it seems probably worth it 😄 

If you feel strongly about it I could pass in the symbol as an argument or something to avoid the circular imports.

**Comment by ChrisPenner** [comment]
@hojberg  It's slower than I'd like TBH, I've considered doing a hybrid approach of caching the last results, serving completions from that while refreshing in the background for the next time; but that might also be confusing if you try to complete a pull right after a push or something. Open to ideas 😄

File: unison-cli/src/Unison/Auth/Tokens.hs (hunk 5 above)
Comment (aryairani) [reply unison-cli/src/Unison/Auth/Tokens.hs:55]: Ah, I see. Yeah Tokens shouldn't really know about InputPatterns, unless we'd want some application reader that knows. But how does input patterns rely back on the auth client?

Comment (ChrisPenner) [reply unison-cli/src/Unison/Auth/Tokens.hs:55]: InputPatterns contain argument types which contain completions which now depend on the authed-client so they can get completions from share.

**Review by mitchellwrosen — APPROVED** [review]

File: unison-src/transcripts/tab-completion.md (hunk 4 above)
Comment (ChrisPenner) [reply unison-src/transcripts/tab-completion.md]: Okay, I moved this one into `transcripts-manual` and added a note there that they need to be run manually.

**Comment by ChrisPenner** [comment]
Okay, I'm going to merge this as-is since it's better than what we've got, but I'd like to experiment with making this faster, and improving "sub-namespace" completions in the future, I'll make an issue 👍🏼 

https://github.com/unisonweb/unison/issues/3441
